### PR TITLE
executor: add limit implementation for CTEExec

### LINKED ISSUE #24869: Support limit in CTEExec
## Development Task
User can use `limit` to easily terminate the recursive CTE. So no need to worry infinite loop. For example:

`with recursive cte1(c1) as (select 1 union select c1 + 1 from cte1 limit 5 offset 1) select * from cte1;`

### PR DESCRIPTION
**Pull request by guo-shaoge**
### What problem does this PR solve?

Issue Number: close #24869 

Problem Summary: 
1. This PR add `limit` implementation in `CTEExec`. The logic is basically same as `LimitExec`.
2. Why add `limit`: in recursive CTE, we can use limit to prevent infinite recursion. Just add limit, the execution will be terminated when it reachs the limitation.
3. Why don't just use `LimitExec`: Because CTEExex is a blocking operator, all data if computed in the first called Next().

### What is changed and how it works?

What's Changed: 
1. Also build limit when building recursive CTE.
2. Add begin info and end info into CTEExec.
3. Add more logic in CTEExec. Basically same as `LimitExec`.

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test: TestCTEWithLimit

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- executor: add limit implementation for CTEExec

### CODE REVIEW

**Comment by guo-shaoge** [comment]
> looks like it includes #24792

We can review [CTEExec](https://github.com/pingcap/tidb/pull/24809) first.

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -432,6 +432,7 @@ type cteInfo struct {
 	// enterSubquery and recursiveRef are used to check "recursive table must be referenced only once, and not in any subquery".
 	enterSubquery bool
 	recursiveRef  bool
+	limitLP       LogicalPlan
```
Comment (wjhuang2016) [inline planner/core/planbuilder.go:435]: Do we really need limitLP?

Comment (guo-shaoge) [reply planner/core/planbuilder.go:435]: limitLP will be tableDual if SQL is `limit 0`

Comment (wjhuang2016) [reply planner/core/planbuilder.go:435]: ok

**Review by xiongjiwei — APPROVED** [review]

**Review by wjhuang2016 — COMMENTED** [review]
explain doesn't show any info about the limit

File: executor/cte_test.go [hunk 2]
```diff
@@ -242,3 +242,65 @@ func (test *CTETestSuite) TestCTEMaxRecursionDepth(c *check.C) {
 	rows = tk.MustQuery("with cte1(c1) as (select 1 union select 2) select * from cte1 order by c1;")
 	rows.Check(testkit.Rows("1", "2"))
 }
+
+func (test *CTETestSuite) TestCTEWithLimit(c *check.C) {
+	tk := testkit.NewTestKit(c, test.store)
+	tk.MustExec("use test;")
+
+	// Basic recursive tests.
+	rows := tk.MustQuery("with recursive cte1(c1) as (select 1 union select c1 + 1 from cte1 limit 5 offset 0) select * from cte1")
```
Comment (xiongjiwei) [inline executor/cte_test.go:251]: can you please add a test like `with recursive cte1(c1) as (select 1 union select c1 + 1 from cte1 limit 5 offset 6) select * from cte1;` I test it in MySQL, it is different from this PR

**Comment by xiongjiwei** [comment]
> explain doesn't show any info about the limit

how about show the limit info like this
image

**Comment by wjhuang2016** [comment]
> > explain doesn't show any info about the limit
> 
> how about show the limit info like this
> image

We can refer to:
```
// ExplainInfo implements Plan interface.
func (p *PhysicalIndexLookUpReader) ExplainInfo() string {
	// The children can be inferred by the relation symbol.
	if p.PushedLimit != nil {
		return fmt.Sprintf("limit embedded(offset:%v, count:%v)", p.PushedLimit.Offset, p.PushedLimit.Count)
	}
	return ""
}
```

File: executor/cte_test.go (hunk 2 above)
Comment (guo-shaoge) [reply executor/cte_test.go:251]: done

**Comment by guo-shaoge** [comment]
> > > explain doesn't show any info about the limit
> > 
> > 
> > how about show the limit info like this
> > image
> 
> We can refer to:
> 
> ```
> // ExplainInfo implements Plan interface.
> func (p *PhysicalIndexLookUpReader) ExplainInfo() string {
> 	// The children can be inferred by the relation symbol.
> 	if p.PushedLimit != nil {
> 		return fmt.Sprintf("limit embedded(offset:%v, count:%v)", p.PushedLimit.Offset, p.PushedLimit.Count)
> 	}
> 	return ""
> }
> ```

done. Check explain_cte.result.

File: executor/cte.go [hunk 3]
```diff
@@ -248,6 +265,9 @@ func (e *CTEExec) computeRecursivePart(ctx context.Context) (err error) {
 			if err = e.setupTblsForNewIteration(); err != nil {
 				return err
 			}
+			if e.hasLimit && uint64(e.resTbl.NumRows()) >= e.limitEnd {
```
Comment (wjhuang2016) [inline executor/cte.go:268]: Better use a function to wrap e.hasLimit && uint64(e.resTbl.NumRows()) >= e.limitEnd

File: cmd/explaintest/t/explain_cte.test [hunk 4]
```diff
@@ -29,3 +29,16 @@ explain with q(a,b) as (select * from t1) select /*+ merge(q) no_merge(q1) */ *
...
+explain with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 c1 from cte1 limit 100 offset 100) select * from cte1;
+explain with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 c1 from cte1 limit 0 offset 0) select * from cte1;
+
+explain with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 c1 from cte1 limit 1) select * from cte1 dt1 join cte1 dt2 on dt1.c1 = dt2.c1;
+explain with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 c1 from cte1 limit 0 offset 0) select * from cte1 dt1 join cte1 dt2 on dt1.c1 = dt2.c1;
+
+# non-recursive limit
+explain with recursive cte1(c1) as (select c1 from t1 union select c1 from t2 limit 1) select * from cte1;
+explain with recursive cte1(c1) as (select c1 from t1 union select c1 from t2 limit 100 offset 100) select * from cte1;
+explain with recursive cte1(c1) as (select c1 from t1 union select c1 from t2 limit 0 offset 0) select * from cte1;
```
Comment (wjhuang2016) [inline cmd/explaintest/t/explain_cte.test:44]: Need tests for non recursive CTE with limit

File: planner/core/physical_plans.go [hunk 5]
```diff
@@ -1463,10 +1463,16 @@ type CTEDefinition PhysicalCTE
 
 // ExplainInfo overrides the ExplainInfo
 func (p *CTEDefinition) ExplainInfo() string {
+	var res string
 	if p.RecurPlan != nil {
-		return "Recursive CTE"
+		res = "Recursive CTE"
+	} else {
+		res = "None Recursive CTE"
```
Comment (wjhuang2016) [inline planner/core/physical_plans.go:1470]: s/None Recursive/Non-recursive

File: executor/cte.go (hunk 3 above)
Comment (guo-shaoge) [reply executor/cte.go]: Done. use `limitDone()` function to check.

File: planner/core/physical_plans.go (hunk 5 above)
Comment (guo-shaoge) [reply planner/core/physical_plans.go]: done

File: cmd/explaintest/t/explain_cte.test (hunk 4 above)
Comment (guo-shaoge) [reply cmd/explaintest/t/explain_cte.test:44]: Alread added in TestWithLimit.

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**